draft-ietf-masque-h3-datagram.md

```diff
@@ -117,26 +123,192 @@ following format (using the notation from the "Notational Conventions" section
 of {{QUIC}}):
 
 ~~~
-HTTP/3 DATAGRAM Frame {
-  Flow ID (i),
+HTTP/3 Datagram {
+  Quarter Stream ID (i),
+  Context ID (i),
   HTTP/3 Datagram Payload (..),
 }
 ~~~
 {: #h3-datagram-format title="HTTP/3 DATAGRAM Frame Format"}
 
-Flow ID:
+Quarter Stream ID:
+
+: A variable-length integer that contains the value of the client-initiated
+bidirectional stream that this datagram is associated with, divided by four.
+(The division by four stems from the fact that HTTP requests are sent on
+client-initiated bidirectional streams, and those have stream IDs that are
+divisible by four.)
+
+Context ID:
+
+: A variable-length integer indicating the context ID of the datagram (see
+{{datagram-contexts}}).
+
+HTTP/3 Datagram Payload:
+
+: The payload of the datagram, whose semantics are defined by individual
+applications. Note that this field can be empty.
+
+Intermediaries parse the Quarter Stream ID field in order to associate the QUIC
+DATAGRAM frame with a stream. If an intermediary receives a QUIC DATAGRAM frame
+whose payload is too short to allow parsing the Quarter Stream ID field, the
+intermediary MUST treat it as an HTTP/3 connection error of type
+H3_GENERAL_PROTOCOL_ERROR. Intermediaries MUST ignore any HTTP/3 Datagram
+fields after the Quarter Stream ID.
+
+Endpoints parse both the Quarter Stream ID field and the Context ID field in
+order to associate the QUIC DATAGRAM frame with a stream and context within
+that stream. If an endpoint receives a QUIC DATAGRAM frame whose payload is too
+short to allow parsing the Quarter Stream ID field, the endpoint MUST treat it
+as an HTTP/3 connection error of type H3_GENERAL_PROTOCOL_ERROR. If an endpoint
+receives a QUIC DATAGRAM frame whose payload is long enough to allow parsing
+the Quarter Stream ID field but too short to allow parsing the Context ID
+field, the endpoint MUST abruptly terminate the corresponding stream with a
+stream error of type H3_GENERAL_PROTOCOL_ERROR.
+
+
+# CAPSULE HTTP/3 Frame Definition {#capsule-frame}
+
+CAPSULE allows reliably sending request-related information end-to-end, even in
+the presence of HTTP intermediaries.
+
+CAPSULE is an HTTP/3 Frame (as opposed to a QUIC frame) which SHALL only be sent
+in client-initiated bidirectional streams. Intermediaries MUST forward all
+received CAPSULE frame in their unmodified entirety on the same stream where it
```

Comment:
Capsules are by definition end-to-end. If intermediaries start looking inside, they might cause ossification. In theory, I'd happily use end-to-end encryption to protect them but we don't have keys for this purpose.

Reply:
I think adding a secure integrity mechanism is overkill. I'd treat these more like how headers and intermediaries work. Most intermediaries really have no business looking at the stuff, but some will decide to do stuff and if they break things then hopefully the failure to abide the spec will be reflected in market position.

Reply:
I was being a bit tongue in cheek by suggesting security measures here. It strikes me as overkill too.

So is it explicitly forbidden for a proxy to register its own context ID on a stream via CAPSULE?  This would require some careful context-ID allocations by the proxy to avoid potential collisions, which in turn break the rules about the context-IDs being end-to-end.

I don't have a use case, but I'm not sure I see the the value in locking this down quite so tightly.  Are there designs that depend on the context identifier being the same at every hop?  Perhaps tunneling QUIC in 'forward' mode relies on this?

Reply:
The core tenet of the two-layer design is that Stream IDs are per-hop and context IDs are end-to-end. That allows extensions to rely on the fact that both endpoints have the same view of context IDs. CAPSULE exists as an end-to-end mechanism, and similarly relies on that property. If we need a new per-hop mechanism, then we can use a different h3 frame. I've clarified that intermediaries don't send capsules apart from forwarding them, so they're not allowed to inject their own context IDs.